File: iOS/examples/Logic3.swift
//
//  Logic3.swift
//  testSDK
//
//  Created by CloudBackend on 2023-03-16.
//

import Foundation

class Logic3: NSObject,
              LogInDelegatePtr, QueryDelegatePtr,
              CreateContainerDelegatePtr, RemoveContainerDelegatePtr,
              CreateObjectDelegatePtr
{
    // --- Global Variables ---
    var cbe1: CloudBackend!
    var rootContainer: CBEContainer!
    var newContainer: CBEContainer!
    var qR: CBEQueryResult!

    // Since the asynchronus code lives in the wrapped C++ code locks are used
    // to ensure that no race conditions are happening.
    let lock = NSLock()
    
    // --- Functions ---
    func waitForRsp() -> Void {
        lock.lock()
    }
    /* Functions for Exercise 2 */
    func removeContainer() {
        var count:Int = 0
        let items:[CBEItem] = qR.getItemsSnapshot() as! [CBEItem]
        while count < items.count {
            let item:CBEItem = items[count] //as! CBEItem
            if item.name() == "newTestContainer" {
                let c = item as! CBEContainer
                c.remove(self)
                waitForRsp()
            }
            count+=1
        }
    }
    func createContainer(container: CBEContainer) -> Void  {
        print("---Creating container")
        let containerName = "newTestContainer"
        rootContainer.createContainer(containerName, delegate:self)
    }
    func loadContainerContents(container: CBEContainer) {
        print("---Loading container contents")
        let filter1 = CBEFilter()
        filter1.setDataType(CBEItemTypeContainer)
        container.query(filter1, delegate:self)
    }
    func printContainerContents(qR: CBEQueryResult) -> Void {
        print("---Printing container contents")
        let items:[CBEItem] = qR.getItemsSnapshot() as! [CBEItem]
        var count:Int = 0
        while count < items.count {
            let item:CBEItem = items[count] //as! CBEItem
            print("testSDK item  \(count) \(String(describing: item)) \(String(describing: item.name()))")
            count+=1
        }
        lock.unlock()
    }
    /* Functions for Exercise 3 */
    func loadContainerObjects(container: CBEContainer) {
        print("---Loading container objects")
        let filter1 = CBEFilter()
        filter1.setDataType(CBEItemTypeObject)
        container.query(filter1, delegate:self)
    }
    
    func printObjects(qR: CBEQueryResult) -> Void {
        print("---Printing objects")
        let items:[CBEItem] = qR.getItemsSnapshot() as! [CBEItem]
        var count:Int = 0
        print(items.count)
        while count < items.count {
            let item:CBEItem = items[count] //as! CBEItem
            print("testSDK item  \(count) \(String(describing: item)) \(String(describing: item.name()))")
            count+=1
        }
        lock.unlock()
    }
    
    func createObject(parentContainer: CBEContainer) -> Void {
        print("---Creating object")
        var metaPairs = Dictionary<String, Dictionary<String, Bool>>()
        var metaData = Dictionary<String, Bool>()

        //add dictionary values - constant time O(1)
        metaData["Volvo"] = false;
        metaPairs["brand"] = metaData;
        parentContainer.createObject("newObject", delegate: self, keyValues: metaPairs)
    }
    
    // --- Success Callbacks ---
    func onLog(inSuccess cloudBackend: CloudBackend!) {
        NSLog("LogIn success ")
        cbe1 = cloudBackend
        rootContainer = cbe1.account().rootContainer()
        lock.unlock()
    }
    func onQuerySuccess(_ queryResult: CBEQueryResult!) {
        qR = queryResult
        lock.unlock()
    }
    func onCreateContainerSuccess(_ container: CBEContainer!) {
        newContainer = container
        lock.unlock()
    }
    func onCreateObjectSuccess(_ object: CBEObject!) {
        lock.unlock()
    }
    func onRemoveContainerSuccess(_ containerId: CBEItemId, name: String!) {
        print("Removed old container...")
        lock.unlock()
    }

    // --- Error Callbacks ---
    func onLog(inError error: CBEError!, context: CBEContext!) {
        NSLog("LogIn error ")
        if let a = error, let b = context {
            NSLog("ErrorCode: %u, reason: %@, message: %@", a.errorCode, a.reason, a.message)
            NSLog("Function: %@, reason: %@", b.fnName, b.report)
        }
    }
    func onQueryError(_ error: CBEError!, context: CBEContext!) {
        if let a = error, let b = context {
            NSLog("ErrorCode: %u, reason: %@, message: %@", a.errorCode, a.reason, a.message)
            NSLog("Function: %@, reason: %@", b.fnName, b.report)
        }
    }
    func onCreateContainerError(_ error: CBEError!, context: CBEContext!) {
        if let a = error, let b = context {
            NSLog("ErrorCode: %u, reason: %@, message: %@", a.errorCode, a.reason, a.message)
            NSLog("Function: %@, reason: %@", b.fnName, b.report)
        }
    }
    func onCreateObjectError(_ error: CBEError!, context: CBEContext!) {
        if let a = error, let b = context {
            NSLog("ErrorCode: %u, reason: %@, message: %@", a.errorCode, a.reason, a.message)
            NSLog("Function: %@, reason: %@", b.fnName, b.report)
        }
    }
    func onRemoveContainerError(_ error: CBEError!, context: CBEContext!) {
        if let a = error, let b = context {
            NSLog("ErrorCode: %u, reason: %@, message: %@", a.errorCode, a.reason, a.message)
            NSLog("Function: %@, reason: %@", b.fnName, b.report)
        }
    }
    
    // --- Driver Code ---
    func runTests()  {
        lock.lock()
        /* Exercise 1 */
        let myUsername : String = "~~~"
        let myPassword : String = "***"
        let myTenant : String = "~~~"
        // Login
        CloudBackend.login(myUsername, password:myPassword, tenant:myTenant, delegate:self)
        waitForRsp()
        NSLog("Logged in as: %@ SDK version %@", cbe1.account().username(), cbe1.version())
        
        /* Exercise 2 */
        // Load container contents ie. query it to get its contents
        loadContainerContents(container: rootContainer)
        waitForRsp()
        
        // Remove container if it already exists
        removeContainer()
        
        // After loading the container, print its contents
        printContainerContents(qR:qR)
        waitForRsp()
        // Given the existing containers, should we create a new?
        createContainer(container:rootContainer)
        waitForRsp()
        
        /* Exercise 3 */
        // Load container contents ie. query it to get its contents
        loadContainerContents(container: rootContainer)
        waitForRsp()
        
        // After loading the container, print its contents
        printContainerContents(qR:qR)

        // Create object in newly created container
        createObject(parentContainer:newContainer)
        waitForRsp()
        
        // Query root to get all objects
        loadContainerObjects(container:rootContainer)
        waitForRsp()
        
        // Print objects contents
        printContainerContents(qR: qR)
        
        NSLog("Exercise Completed!")
    }
}

File: C++/Examples/TutorialAsyncDelegates/Logic3.cpp
#include "Logic.h"

#include <algorithm>  // std::find_if_not
#include <cctype>     // std:isspace, std::tolower, std::toupper
#include <cstring>
#include <iomanip>    // std::setw
#include <ios>        // std::left
#include <sstream>    // std::ostringstream
#include <stdexcept>
#include <string>

#include "MyDelegates.cpp"
#include "MyGenericFunctions.cpp"

// - - - - - - - - - - - - - - - - - FUNCTIONS - - - - - - - - - - - - - - - - - 

// Exit function that reports how program was terminated
void Exercise::exitProgram(const int exitCode) {
  std::ostringstream oss;
  oss << "Exiting program with exit code: " << exitCode << ".";
  if (myCloudBackend) {
    myCloudBackend.terminate();
  }
  throw std::invalid_argument( oss.str() );
}

/* Exercise 2 */

// Creates a new container in the parentContainer argument that is passed in
void Exercise::createContainer(cbe::Container parentContainer) {
  std::cout << "Logic: Create container" << std::endl;
  // Generic function to get string from user input
  const auto name = inquireString("Set name for Container");

  // Create a delegate that is going to be used for the createContainer method
  std::shared_ptr<MyCreateContainerDelegate> createContainerDelegate = 
                                  std::make_shared<MyCreateContainerDelegate>();
  parentContainer.createContainer(name, createContainerDelegate);
  
  // Calling wait for the delegate to finish
  createContainerDelegate->waitForRsp();

  // Check if error, can also check if a container was constructed.
  if (createContainerDelegate->errorInfo){ 
    std::cout << "Error!" << std::endl;
    std::cout << "Container creation failed." << std::endl;
    
    // Print info about error 
    std::cout << createContainerDelegate->errorInfo << std::endl;
    
    // Exit program with unique code 
    exitProgram(21);
  }

  // Get container from the delegate
  this->myContainer = createContainerDelegate->container;
}

// Loads container contents to be able to see what is inside the container that 
// is passed in
void Exercise::loadContainerContents(cbe::Container container) {
  // Generic function to print information about container
  std::cout << "Getting sub-container of " << containerName(container) 
            << std::endl;
  
  // Create a filter that limits the query to only return containers 
  cbe::Filter containerFilter;
  containerFilter.setDataType(cbe::ItemType::Container);
  std::shared_ptr<MyQueryDelegate> queryDelegate = 
                                            std::make_shared<MyQueryDelegate>();

  // Call query function with filter and delegate
  container.query(containerFilter, queryDelegate);

  // Wait for delegate to finish
  queryDelegate->waitForRsp();
  
  // Check if error 
  if (queryDelegate->errorInfo){ // Could also check if qResult was constructed
    std::cout << "Error!" << std::endl;
    std::cout << "Could not load container contents." << std::endl;

    // Print info about error 
    std::cout << queryDelegate->errorInfo << std::endl;

    // Exit program with unique code
    exitProgram(30);
  }
  
  // Get queryResult from the delegate
  this->qResult = queryDelegate->queryResult;
}

// Function that prints container contents
void Exercise::printContainerContents(cbe::QueryResult qResult) {  
  // Put queryResult in vector of cbe::Item 
  std::vector<cbe::Item> items = qResult.getItemsSnapshot();
  std::cout << "------------------------------------------------------------" 
            << std::endl;
  // Loop through items and print them
  for (const auto& item : items) {
    // Generic print function 
    printItem(item);
  }
  std::cout << "------------------------------------------------------------" 
            << std::endl;
}

/* Exercise 3 */

// The same as loadContainerContents but with different filter
void Exercise::loadContainerObjects(cbe::Container container) {
  std::cout << "Getting objects in container " << containerName(container) 
            << std::endl;
  std::shared_ptr<MyQueryDelegate> queryDelegate = 
                                            std::make_shared<MyQueryDelegate>();

  // Create a filter that limits the query to only return objects 
  cbe::Filter objectFilter;
  objectFilter.setDataType(cbe::ItemType::Object);
  
  // Call query function with filter and delegate
  container.query(objectFilter, queryDelegate);

  // Wait for delegate to finish
  queryDelegate->waitForRsp();
   // Check if error 
  if (queryDelegate->errorInfo){ // Can check if qResult was constructed.
    std::cout << "Error!" << std::endl;
    std::cout << "Could not load container contents." << std::endl;

    // Print info about fail
    std::cout << queryDelegate->errorInfo << std::endl;

    // Exit program with unique code 
    exitProgram(33);
  }

  // Get queryResult from the delegate
  this->qResult = queryDelegate->queryResult;
}

void Exercise::printObjects(cbe::QueryResult queryResult) {
  cbe::Object tempObject{cbe::DefaultCtor{}}; // nullptr;
  std::cout << "Printing Objects from query result: " << std::endl;
  std::cout << "------------------------------------------------------------" 
            << std::endl;
  
  // Loop through objects from queryResult
  for (const auto& item : queryResult.getItemsSnapshot()) {
    if (item.type() == cbe::ItemType::Object) {
      printItem(item);
      tempObject = cbe::CloudBackend::castObject(item);

      // Get key/value information to print
      const auto keyValues = tempObject.keyValues();
      if (!keyValues.empty()) {

        // Loop through key/values for each object
        for (const auto& keyValue : keyValues) {
          const auto& key      = keyValue.first;
          const auto& sdkValue = keyValue.second;
          const auto  indexed  = std::get<bool>(sdkValue);
          std::cout << "    " << key << " = " << "'"
                    << std::get<std::string>(sdkValue)
                    << "'" << (indexed? " \t\t(indexed)" : "") << std::endl;
        }
      }  // if (!keyValues.empty())
    }   // if (item.type() == cbe::ItemType::Object)
  }
  std::cout << "------------------------------------------------------------" 
            << std::endl;

  // If container does not contain any objects
  if (!tempObject) {
    std::cout << "Sorry, no objects found in the Container!" << std::endl;
  }
}

cbe::Container Exercise::selectContainer(const std::string& prompt) {
  std::cout << "Select Container" << std::endl;
  const auto items = this->qResult.getItemsSnapshot();
  
  // Do the following until user has entered a container that exists 
  while (true) {
    // Generic function to get string from user input
    const auto containerName = inquireString(prompt);

    // Loop through and look for container that user has requested
    for (const auto& item : items) {

      // If container is found, return it 
      if (item.name() == containerName) {
        return cbe::CloudBackend::castContainer(item);
      }
    }

    // If container is not fount an error is printed
    std::cout << "Error: the container you asked for, \"" << containerName
              << "\", was not found; following are the options:" << std::endl;
    // Loops through available containers to select 
    for (const auto& item : items) {
      printItem(item);
    }
  }
}

cbe::Object Exercise::createObject(cbe::Container inContainer) {
  std::cout << "Create Object" << std::endl;

  // Generic functions to get string from user input
  const std::string name = inquireString("Set name for Object");

  // Set the number of key/values the user wants to input 
  const int numOfTags =
                       inquireInt("Set the number of Key/Value pairs you want");
  
  // Create variable for key/values
  cbe::KeyValues keyValues;

  for (int i = 1; i <= numOfTags; i++) {
    
    // String that will contain the key name 
    std::string keyName{};

    // Set keyName and check if it is valid
    while(true) {
      keyName = inquireString("Name of Key #" + std::to_string(i));
      if(!isdigit(keyName[0])){
        break;
      }
      // Error message if user input starts with a number
      std::cout << "Key can not start with numerical! " << std::endl;
    }

    // Generic function to get string from user input
    const auto value =
        inquireString("Value for #" + std::to_string(i) + " '" + keyName + "'");
    
    // Generic function to get bool from user input
    // Indexed key/values are quicker to access through searching
    const auto indexed =
                       inquireBool("Make KeyValue pair #" + std::to_string(i) +
                                   " indexed or not (y indexed, n not indexed",
                                   true /* defaultVal */);
    // Insert value for key
    keyValues[keyName] = std::pair<std::string, bool>(value, indexed);
  }

  // Delegate for createObject
  std::shared_ptr<MyCreateObjectDelegate> createObjectDelegate = 
                                    std::make_shared<MyCreateObjectDelegate>();  
  
  // Check if user has created any key/values
  if (numOfTags > 0) {
    // Create object with name, delegate and key/values
    inContainer.createObject(name, keyValues, createObjectDelegate);
  } else {
    // Create object without key/values
    inContainer.createObject(name, createObjectDelegate);
  }
  
  // Wait for delegate to finish
  createObjectDelegate->waitForRsp();
  
  // Check if error 
  if (createObjectDelegate->errorInfo){ 
    std::cout << "Error!" << std::endl;
    std::cout << "Could not create object." << std::endl;

    // Print info about error
    std::cout << createObjectDelegate->errorInfo << std::endl;

    // Exit with unique code
    exitProgram(31);
  }
  // Return object from delegate
  return createObjectDelegate->object;    
}

// - - - - - - - - - - - - - - LOGIC EXERCISE 1,2,3 - - - - - - - - - - - - - - 

void Exercise::logic() {
  /* Exercise 1 */
  // Generic function to get bool from user input  
  const bool login = inquireBool("Do you want to login", true /* defaultVal*/);
  
  // If user responds with a no then exit 
  if (!login) {
    // Exit program with unique code
    exitProgram(11);
  }
  // Generic function to get strings from user input
  // Second argument is "default value", user only has to press enter 
  const auto username = inquireString("Type username", "gitHubTester1");
  const auto password = inquireString("Type password");
  const auto tenant   = inquireString("Enter  tenant", "cbe_githubtesters");

  // Create log in delegate
  std::shared_ptr<MyLogInDelegate> logInDelegate = 
                                            std::make_shared<MyLogInDelegate>();
 
  // Call login and set member variable
  myCloudBackend = cbe::CloudBackend::logIn(username, 
                                            password, 
                                            tenant,
                                            logInDelegate);

  // Wait for delegate to finish
  logInDelegate->waitForRsp();
  
  // Check if error 
  if (logInDelegate->errorInfo){
    // Print info about error
    std::cout << "Error, login failed! \nError info=" 
              << logInDelegate->errorInfo << std::endl;
    
    // Exit program with unique code
    exitProgram(10);
  }

  // Confirm login and SDK version
  std::cout << "\nLogged in as:\t" << myCloudBackend.account().username()
            << "\nSDK version:\t"  << myCloudBackend.version() << "\n" 
            << std::endl; 
  
  /* Exercise 2 */
  // Put root container in member variable
  this->rootContainer = logInDelegate->cloudBackend.account().rootContainer();     
  
  // Load container contents ie. query it to get its contents
  loadContainerContents(this->rootContainer);
  
  // After loading the container, print its contents
  printContainerContents(this->qResult);
  
  // Given the existing containers, should we create a new?
  if (inquireBool("Do you want to create a new container", false)) {
    
    // Create a new container in the container referred container
    createContainer(rootContainer); 

    // Print some information about container
    std::cout << "Container name: "
              << containerName(this->myContainer, true /* temporary */)
              << std::endl;
  } 

  /* Exercise 3 */

  // Load container contents ie. query it to get its contents
  loadContainerContents(rootContainer);

  // After loading the container, print its contents
  printContainerContents(this->qResult);

  // Print the number of items in the container
  std::cout << "Item count: " << this->qResult.getItemsSnapshot().size() 
            << std::endl;
  
  // Prompts the user to select a container in which an object is to be created
  this->myContainer = 
     selectContainer("In which Container do you want to create the new object");

  // Print the container the user selected
  std::cout << "Container: " << containerName(this->myContainer) << std::endl;
  
  // Creates an object in the selected container
  object = createObject(this->myContainer);

  // Prints information about the object that has been created
  std::cout << "Designated name: " << objectName(object, true) << std::endl;
  std::cout << "Object name: " << objectName(object) << std::endl;

  // Finally after the object has been created, load its contents
  loadContainerObjects(this->myContainer);

  // Print all objects and its content  
  printObjects(this->qResult);
  
  std::cout << "Exercise completed!" << std::endl;
  myCloudBackend.terminate();
} // void Exercise::logic
